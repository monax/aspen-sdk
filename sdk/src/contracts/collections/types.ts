import { Provider } from '@ethersproject/abstract-provider';
import type { BigNumber, Signer } from 'ethers';
import { Address, ChainId } from '../index.js';
import type { CollectionMetaImageType, CollectionMetaLinkType } from './constants';

export type Signerish = Signer | Provider;

export type TokenStandard = 'ERC721' | 'ERC1155' | 'ERC20';

export type ContractVerificationType = 'aspen-minted' | 'aspen-partner' | 'os-verified';

export type MetadataKind = 'collection';

export type CollectionMetadataPhase = {
  name?: string;
  start_date?: string;
  token_id?: string | number | null;
  banner_button_text?: string | null;
  header_text?: string | null;
};

export type CollectionCreator = {
  creator_name: string;
  creator_email: string;
};

export type CollectionMetadata = {
  schema?: string | null;
  kind?: MetadataKind;
  name: string;
  description?: string;
  alternative_text?: string;
  image?: string;
  categories?: number[];
  creators?: CollectionCreator[];
  external_link?: string;
  explicit_content?: false;
  images?: { [key in keyof typeof CollectionMetaImageType]?: string };
  links?: { [key in keyof typeof CollectionMetaLinkType]?: string };
  aspen?: {
    mint: {
      phases: CollectionMetadataPhase[];
    };
  };
  // the following are only returned via the API
  terms?: {
    url?: string | null;
    required?: boolean | null;
  };
  advert?: {
    name?: string;
    content?: string;
    cta?: string;
    imageUrl?: string;
    redirectUrl?: string;
  };
  advertVideo?: {
    name?: string;
    content?: string;
    cta?: string;
    videoUrl?: string;
    redirectUrl?: string;
  };
};

// NOTE: NftMetadataAttribute looks very loose on requirements
export type TokenMetadataAttribute = {
  trait_type: string;
  value?: unknown;
};

// NOTE the code in Cedar API slightly differs from what's defined in oas/NftMetadata
export type TokenMetadata = {
  name: string;
  description?: string;
  external_url?: string;
  image?: string;
  image_ipfs?: string;
  animation_url?: string;
  animation_url_ipfs?: string;
  attributes: TokenMetadataAttribute[];
  // TODO: media?
};

export type TokenAsset = {
  chainId: ChainId;
  contractAddress: string;
  tokenId: string;
  standard: TokenStandard;
  quantity: BigNumber;
};

export type TokenAssetMetadata = {
  asset: TokenAsset;
  uri: string | null;
  metadata: TokenMetadata | null;
};

export type CollectionContractClaimCondition = {
  startTimestamp: number;
  maxClaimableSupply: BigNumber;
  supplyClaimed: BigNumber;
  quantityLimitPerTransaction: BigNumber;
  waitTimeInSecondsBetweenClaims: number;
  merkleRoot: string;
  pricePerToken: BigNumber;
  currency: Address;
  isClaimingPaused: boolean;
};

export type ActiveClaimConditions = {
  maxWalletClaimCount: BigNumber;
  tokenSupply: BigNumber;
  maxTotalSupply: BigNumber;
  maxAvailableSupply: BigNumber;
  activeClaimConditionId: number;
  activeClaimCondition: CollectionContractClaimCondition;
};

export type UserClaimConditions = {
  activeClaimConditionId: number;
  walletClaimCount: BigNumber;
  walletClaimedCountInPhase: BigNumber | null;
  lastClaimTimestamp: number;
  nextClaimTimestamp: number;
};

export type CollectionUserClaimState =
  | 'ok'
  | 'unknown'
  | 'paused'
  | 'not-allowlisted'
  | 'no-token-supply'
  | 'minting-throttled'
  | 'claimed-allowlist-allowance'
  | 'claimed-phase-allowance'
  | 'claimed-wallet-allowance';

export type CollectionUserClaimConditions = {
  availableQuantity: BigNumber;
  canClaimTokens: boolean;
  canMintAfter: Date;
  claimState: CollectionUserClaimState;
};

export type TermsUserAcceptanceState = {
  termsActivated: boolean;
  termsAccepted: boolean;
  termsLink: string;
  termsVersion: number;
};

export type CollectionInfo = {
  chainId: ChainId | null;
  address: Address;
  tokenStandard: TokenStandard | null;
};

export type CollectionCallData = {
  method: string;
  args?: { [key: string]: unknown };
  signer: Signerish;
  supportedFeatures: string[];
};

export type DebugHandler = (collection: CollectionInfo, message: string, ...optionalParams: unknown[]) => void;

export type ErrorHandler = (
  message: string,
  error: Error,
  collection: CollectionInfo,
  callData: CollectionCallData,
) => void;
