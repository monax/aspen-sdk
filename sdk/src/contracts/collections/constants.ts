export const TokenStandards = ['ERC721', 'ERC1155'];

export const CollectionMetaImageType = {
  banner: 'banner',
} as const;

export const CollectionMetaLinkType = {
  website: 'website',
  twitter: 'twitter',
  discord: 'discord',
  instagram: 'instagram',
  telegram: 'telegram',
  linkedin: 'linkedin',
};
