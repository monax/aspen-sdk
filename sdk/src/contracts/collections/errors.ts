export enum SdkErrorCode {
  FAILED_TO_LOAD_FEATURES = 'Failed to load contract features.',
  EMPTY_TOKEN_STANDARD = "Couldn't identify a token standard",
  FEATURE_NOT_SUPPORTED = 'This feature is not supported by the contract',
  FUNCTION_NOT_SUPPORTED = 'Function is not supported by the contract',
  FUNCTION_NOT_IMPLEMENTED = 'Function is not implemented',
  CHAIN_ERROR = 'A call to the chain failed',
  WEB_REQUEST_FAILED = 'Web request failed!',
  TOKEN_ID_REQUIRED = 'Token Id is required to call the function on this contract!',
  TOKEN_ID_REJECTED = 'Token Id is not supported for this function on this contract!',
  INVALID_DATA = 'Invalid input data provided!',
  UNKNOWN_ERROR = 'Unknown error.',
}

export class SdkError extends Error {
  constructor(
    code: SdkErrorCode,
    public readonly data: Record<string, unknown> = {},
    public readonly error: Error | null = null,
  ) {
    super(code);

    this.name = 'SdkError';
  }

  static is(err: unknown): err is SdkError {
    return typeof err === 'object' && (err as SdkError).name === 'SdkError';
  }

  static from(err: unknown, code: SdkErrorCode, data?: Record<string, unknown>): SdkError {
    // TODO - parse the error if code == SdkErrorCode.CHAIN_ERROR
    return SdkError.is(err) ? err : new SdkError(code, data, err as Error);
  }
}
