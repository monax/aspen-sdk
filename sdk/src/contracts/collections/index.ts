export * from './collections';
export * from './constants';
export * from './errors';
export * from './features';
export * from './objects';
export * from './types';
