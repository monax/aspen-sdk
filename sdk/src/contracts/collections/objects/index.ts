export * from './claim';
export * from './issue';
export * from './token';
