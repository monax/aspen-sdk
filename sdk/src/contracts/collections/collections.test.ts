import { describe, expect, test } from '@jest/globals';
import { parse } from '@monaxlabs/phloem/dist/schema';
import { Address } from '@monaxlabs/phloem/dist/types';
import { BigNumber } from 'ethers';
import { CollectionContract } from './collections';
import { SdkErrorCode } from './errors';
import {
  asExecutable,
  ContractFunctionId,
  ERC1155StandardInterfaces,
  ERC721StandardInterfaces,
  ExperimentalFeatureInterfaceId,
  FeatureFunctionId,
  FeatureInterfaceId,
  HasAcceptedTerms,
  hasAcceptedTerms,
} from './features';
import { ExperimentalFeatures } from './features/experimental-features.gen';
import { FeatureFactories } from './features/feature-factories.gen';
import { FeatureFunctionsMap } from './features/feature-functions.gen';
import { Token } from './objects';
import { MockJsonRpcProvider } from './utils.testing';

const CONTRACT_ADDRESS = parse(Address, '0xB2Af02eC55E2ba5afe246Ed51b8aBdBBa5F7937C');

const NON_DROP_CONTRACT_INTERFACE_FILES = [
  // Standards
  'standard/IERC20.sol',
  'standard/IERC4906.sol',
  // Payment Notary
  'payments/IPaymentNotary.sol',
  'subscriptions/IPaymentNotary.sol',
  // Core Registry
  'registry/ICoreRegistry.sol',
  // Config
  'config/IGlobalConfig.sol',
  'config/ITieredPricing.sol',
  'config/IOperatorFilterersConfig.sol',
  'registry/ICoreRegistry.sol',
  // Error interfaces
  'errors/ICoreErrors.sol',
  'errors/IDropErrors.sol',
  'errors/IPaymentsErrors.sol',
  'errors/ISplitPaymentErrors.sol',
  'errors/ITermsErrors.sol',
  'errors/IUUPSUpgradeableErrors.sol',
  'errors/IAspenFactoryErrors.sol',
  // Other
  'pausable/ICedarPausable.sol',
  'splitpayment/ISplitPayment.sol',
  'splitpayment/ICedarSplitPayment.sol',
  'issuance/ICedarPremint.sol',
  'issuance/ICedarERC20Payable.sol',
  'issuance/ICedarIssuance.sol',
  'issuance/ICedarOrderFiller.sol',
  'issuance/ICedarNativePayable.sol',
  'issuance/ICedarClaimable.sol',
  'agreement/IAgreementsRegistry.sol',
  'baseURI/ICedarUpgradeBaseURI.sol',
];

// The following interfaces don't relate to any drop contract and so they don't need implementing
const NON_DROP_CONTRACT_INTERFACES: FeatureInterfaceId[] = ['ownable/IOwnable.sol:IOwnableEventV0'];

const NOT_IMPLEMENTED_FUNCTIONS: FeatureFunctionId[] = [
  // Note: add here any unimplemented functions that will be implemented later.
];

const isDropInterface = (iface: FeatureInterfaceId, allowExperimental: boolean): boolean => {
  const file = iface.split(/[:]/).shift() as string;
  const isExperimental = ExperimentalFeatures.includes(iface as ExperimentalFeatureInterfaceId);
  return !NON_DROP_CONTRACT_INTERFACE_FILES.includes(file) && (allowExperimental || !isExperimental);
};

const isDropFunction = (func: FeatureFunctionId, allowExperimental: boolean): boolean => {
  const interfaces = Object.values(FeatureFunctionsMap[func].drop);
  return interfaces.filter((f) => isDropInterface(f, allowExperimental)).length !== 0;
};

describe('Collections - static tests', () => {
  test('Check token standard interface matchers', () => {
    const combinedActual = [...ERC721StandardInterfaces, ...ERC1155StandardInterfaces];
    const combinedExpected = [
      ...FeatureFunctionsMap['isApprovedForAll(address,address)[bool]'].drop,
      ...FeatureFunctionsMap['exists(uint256)[bool]'].drop,
      ...FeatureFunctionsMap['totalSupply()[uint256]'].drop,
      ...FeatureFunctionsMap['totalSupply(uint256)[uint256]'].drop,
    ].filter((id) => /(ERC721|NFT|ERC1155|SFT)/.test(id));

    expect(Array.from(new Set(combinedActual)).sort()).toStrictEqual(Array.from(new Set(combinedExpected)).sort());

    const overlap = ERC721StandardInterfaces.filter((f) => ERC1155StandardInterfaces.includes(f));
    expect(overlap).toStrictEqual([]);
  });

  test('Check that all released functions are implemented', () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0']);

    const allImplementedFunctions: FeatureFunctionId[] = Object.values(erc721.getFunctionsProps('handledFunctions'))
      .map(Object.values)
      .flat();

    const missingFunctions = (Object.keys(FeatureFunctionsMap) as FeatureFunctionId[]).filter(
      (f) =>
        isDropFunction(f, false) && !(allImplementedFunctions.includes(f) || NOT_IMPLEMENTED_FUNCTIONS.includes(f)),
    );

    // immediately show what's not implemened
    expect(missingFunctions).toStrictEqual([]);
  });

  test('Check that all experimental functions are implemented', () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0']);

    const allImplementedFunctions: FeatureFunctionId[] = Object.values(erc721.getFunctionsProps('handledFunctions'))
      .map(Object.values)
      .flat();

    const missingFunctions = (Object.keys(FeatureFunctionsMap) as FeatureFunctionId[]).filter(
      (f) => isDropFunction(f, true) && !(allImplementedFunctions.includes(f) || NOT_IMPLEMENTED_FUNCTIONS.includes(f)),
    );

    // We endeavour to implement all experimental features, but if we choose to deviate make it an explicit decision
    // to update snapshot
    expect(missingFunctions).toMatchSnapshot();
  });

  test('Check that every function lists at least the interfaces of the functions it states to implemenst', () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0']);

    const allImplementedFeatures = erc721.getFunctionsProps('handledFeatures');

    Object.entries(erc721.getFunctionsProps('handledFunctions')).map(([func, handledFunctions]) => {
      const expectedInterfaces = Object.values(handledFunctions)
        .map((f) => FeatureFunctionsMap[f].drop)
        .flat();
      const listedInterfaces = allImplementedFeatures[func as ContractFunctionId];
      const missingInterfaces = expectedInterfaces.filter((i) => !listedInterfaces.includes(i));
      // We endeavour to implement all experimental features, but if we choose to deviate make it an explicit decision
      // to update snapshot
      expect(missingInterfaces).toMatchSnapshot();
    });
  });

  test('Check that all released features are implemented', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0']);

    const allImplementedFeatures = Object.values(erc721.getFunctionsProps('handledFeatures')).flat();

    const missingFeatures = (Object.keys(FeatureFactories) as FeatureInterfaceId[]).filter(
      (f) =>
        isDropInterface(f, false) && !(allImplementedFeatures.includes(f) || NON_DROP_CONTRACT_INTERFACES.includes(f)),
    );

    // We endeavour to implement all experimental features, but if we choose to deviate make it an explicit decision
    // to update snapshot
    expect(missingFeatures).toMatchSnapshot();
  });

  test('Check that all experimental features are implemented', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0']);

    const allImplementedFeatures = Object.values(erc721.getFunctionsProps('handledFeatures')).flat();

    const missingFeatures = (Object.keys(FeatureFactories) as FeatureInterfaceId[]).filter(
      (f) =>
        isDropInterface(f, true) && !(allImplementedFeatures.includes(f) || NON_DROP_CONTRACT_INTERFACES.includes(f)),
    );

    // We endeavour to implement all experimental features, but if we choose to deviate make it an explicit decision
    // to update snapshot
    expect(missingFeatures).toMatchSnapshot();
  });

  test('Experimental flag', async () => {
    const provider = new MockJsonRpcProvider();
    const regular = ['standard/IERC721.sol:IERC721V0'];
    const combinedInterfaces = [...regular, ...ExperimentalFeatures];

    const contractRegular = new CollectionContract(provider, 1, CONTRACT_ADDRESS, combinedInterfaces);
    expect(contractRegular.supportedFeaturesList).toStrictEqual([...regular]);

    const contractExperimental = new CollectionContract(provider, 1, CONTRACT_ADDRESS, combinedInterfaces, true);
    expect(contractExperimental.supportedFeaturesList).toStrictEqual(combinedInterfaces);
  });

  test('Token standard detection', async () => {
    const provider = new MockJsonRpcProvider();

    // contract expects at least one token standard to be supported
    expect(() => new CollectionContract(provider, 1, CONTRACT_ADDRESS, [])).toThrow(SdkErrorCode.EMPTY_TOKEN_STANDARD);

    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V0', 'xxx']);
    expect(erc721.tokenStandard).toBe('ERC721');
    expect(erc721.supportedFeaturesList).toStrictEqual(['standard/IERC721.sol:IERC721V0']);
    expect(async () => await new Token(erc721, 0).exists()).rejects.toThrow(SdkErrorCode.CHAIN_ERROR);

    const erc1155 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC1155.sol:IERC1155V0', 'yyy']);
    expect(erc1155.tokenStandard).toBe('ERC1155');
    expect(erc1155.supportedFeaturesList).toStrictEqual(['standard/IERC1155.sol:IERC1155V0']);
  });

  test('ERC721 Token existence & supply', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, [
      'standard/IERC721.sol:IERC721V0',
      'issuance/INFTSupply.sol:INFTSupplyV0',
    ]);

    expect(erc721.totalSupply.supported).toBe(true);

    const iface = erc721.assumeFeature('issuance/INFTSupply.sol:INFTSupplyV0').interface;

    const exists = iface.encodeFunctionResult(iface.functions['exists(uint256)'], [true]);
    provider.addMock('call', exists);
    expect(await new Token(erc721, 0).exists()).toBe(true);

    // for ERC721 the token supply is either 1 or 0 depending on its existence
    const doesntExist = iface.encodeFunctionResult(iface.functions['exists(uint256)'], [false]);
    provider.addMock('call', doesntExist);
    expect((await new Token(erc721, null).totalSupply()).toNumber()).toBe(0);
  });

  test('ERC1155 Token existence & supply', async () => {
    const provider = new MockJsonRpcProvider();
    const erc1155 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, [
      'standard/IERC1155.sol:IERC1155V1',
      'standard/IERC1155.sol:IERC1155SupplyV2',
    ]);

    expect(erc1155.totalSupply.supported).toBe(true);

    const iface = erc1155.assumeFeature('standard/IERC1155.sol:IERC1155SupplyV2').interface;

    const exists = iface.encodeFunctionResult(iface.functions['exists(uint256)'], [true]);
    provider.addMock('call', exists);
    expect(await new Token(erc1155, 0).exists()).toBe(true);

    const supply = iface.encodeFunctionResult(iface.functions['totalSupply(uint256)'], [BigNumber.from(20)]);
    provider.addMock('call', supply);
    expect((await new Token(erc1155, 0).totalSupply()).toNumber()).toBe(20);
  });

  test('ERC721 Token metadata', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, [
      'standard/IERC721.sol:IERC721V0',
      // Some of our old contracts didn't correctly support metadata interface,
      // so even if the contract doesn't explicitly support the interface
      // we should still be able to call 'tokenURI' function and get a result.
      // 'metadata/INFTMetadata.sol:IAspenNFTMetadataV1'
    ]);

    const ipfsUri = 'ipfs://QmWU5iKU65xY7bxYQgNLrmEBc3MRr5p8owmdFuk94XXTwS/0';
    const iface = erc721.assumeFeature('metadata/INFTMetadata.sol:IAspenNFTMetadataV1').interface;
    const supply = iface.encodeFunctionResult(iface.functions['tokenURI(uint256)'], [ipfsUri]);
    provider.addMock('call', supply);
    expect(await new Token(erc721, 0).getUri()).toBe(ipfsUri);
  });

  test('ERC1155 Token metadata', async () => {
    const provider = new MockJsonRpcProvider();
    const erc1155 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, [
      'standard/IERC1155.sol:IERC1155V1',
      // Some of our old contracts didn't correctly support metadata interface,
      // so even if the contract doesn't explicitly support the interface
      // we should still be able to call 'uri' function and get a result.
      // 'metadata/ISFTMetadata.sol:IAspenSFTMetadataV1',
    ]);

    const ipfsUri = 'ipfs://QmWU5iKU65xY7bxYQgNLrmEBc3MRr5p8owmdFuk94XXTwS/0';
    const iface = erc1155.assumeFeature('metadata/ISFTMetadata.sol:IAspenSFTMetadataV1').interface;
    const supply = iface.encodeFunctionResult(iface.functions['uri(uint256)'], [ipfsUri]);
    provider.addMock('call', supply);
    expect(await new Token(erc1155, 0).getUri()).toBe(ipfsUri);
  });

  test('ERC721 Contract name and symbol', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, ['standard/IERC721.sol:IERC721V2']);

    const contractName = 'Aspen T&Cs Demo';
    const contractSymbol = 'ATD';
    const iface = erc721.assumeFeature('standard/IERC721.sol:IERC721V2').interface;

    const nameResponse = iface.encodeFunctionResult(iface.functions['name()'], [contractName]);
    provider.addMock('call', nameResponse);
    expect(await erc721.name()).toBe(contractName);

    const symbolResponse = iface.encodeFunctionResult(iface.functions['symbol()'], [contractSymbol]);
    provider.addMock('call', symbolResponse);
    expect(await erc721.symbol()).toBe(contractSymbol);
  });

  test('Callable class instance', async () => {
    const provider = new MockJsonRpcProvider();
    const erc721 = new CollectionContract(provider, 1, CONTRACT_ADDRESS, [
      'standard/IERC721.sol:IERC721V0',
      'agreement/IAgreement.sol:IPublicAgreementV1',
    ]);

    const iface = erc721.assumeFeature('agreement/IAgreement.sol:IPublicAgreementV1').interface;
    const doesExist = iface.encodeFunctionResult(iface.functions['hasAcceptedTerms(address)'], [true]);
    provider.addMock('call', doesExist);
    provider.addMock('call', doesExist);
    provider.addMock('call', doesExist);

    const callable = asExecutable(new HasAcceptedTerms(erc721));
    const applyTrap = hasAcceptedTerms(erc721);
    const constructTrap = new hasAcceptedTerms(erc721);
    expect(await callable(CONTRACT_ADDRESS)).toBe(true);
    expect(await applyTrap(CONTRACT_ADDRESS)).toBe(true);
    expect(await constructTrap(CONTRACT_ADDRESS)).toBe(true);
  });
});
