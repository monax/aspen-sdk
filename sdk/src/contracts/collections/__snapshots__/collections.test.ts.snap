// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collections - static tests Check that all experimental features are implemented 1`] = `[]`;

exports[`Collections - static tests Check that all experimental functions are implemented 1`] = `[]`;

exports[`Collections - static tests Check that all released features are implemented 1`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 1`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 2`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 3`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 4`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 5`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 6`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 7`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 8`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 9`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 10`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 11`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 12`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 13`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 14`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 15`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 16`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 17`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 18`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 19`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 20`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 21`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 22`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 23`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 24`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 25`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 26`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 27`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 28`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 29`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 30`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 31`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 32`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 33`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 34`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 35`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 36`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 37`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 38`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 39`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 40`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 41`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 42`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 43`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 44`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 45`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 46`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 47`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 48`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 49`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 50`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 51`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 52`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 53`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 54`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 55`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 56`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 57`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 58`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 59`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 60`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 61`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 62`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 63`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 64`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 65`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 66`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 67`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 68`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 69`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 70`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 71`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 72`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 73`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 74`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 75`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 76`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 77`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 78`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 79`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 80`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 81`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 82`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 83`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 84`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 85`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 86`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 87`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 88`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 89`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 90`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 91`] = `[]`;

exports[`Collections - static tests Check that every function lists at least the interfaces of the functions it states to implemenst 92`] = `[]`;
